### PR TITLE
Enable CudaGraph Annotations

### PR DESCRIPTION
**Pull request by SherlockNoMad**
Human Note: this is for enabling the mark_kernels, so that we see annotations in profiling trace when cudagraph is used. 
Also removed the duplicated implementation in GraphTrainer. 

## Summary

Graph Trainer maintained its own CUDA graph manager and wrapper even though core TorchTitan already provided the same capture, warmup, input-copy, replay, and teardown behavior. The two implementations had begun to diverge, which made CUDA graph fixes and profiler annotation support harder to maintain.

This PR:

- adds explicit `tensor_input_indices` support to the core `CUDAGraphWrapper`
- moves CUDA graph annotation collection and trace post-processing into the core implementation
- always enables CUDA graph annotations and removes the profiler `trace_post_processor` configuration field
- updates Graph Trainer to use the core CUDA graph wrapper and teardown path
- removes the duplicated Graph Trainer CUDA graph manager and wrapper
- temporarily disables `test_precompile_vs_trace`, because loaded precompile artifacts currently provide no example inputs for CUDA graph setup

The precompile test will be fixed and re-enabled in a follow-up by constructing the wrapper with live inputs from the training loop instead of serializing inputs in the artifact.

## Test plan

- `pytest -q tests/unit_tests/cpu/test_cudagraph.py torchtitan/experiments/graph_trainer/tests/test_precompile.py torchtitan/experiments/graph_trainer/tests/test_profiler.py`
- `pytest -q torchtitan/experiments/graph_trainer/tests/test_passes.py -k cudagraph`
- `pytest -q torchtitan/experiments/graph_trainer/tests/test_trace_module.py -k cudagraph`
- `pytest -q torchtitan/experiments/graph_trainer/tests/test_bitwise_deterministic.py::TestLlama3BitwiseDeterministic::test_aot_fx_trace_vs_eager`
- targeted `ufmt` and `flake8` checks

### CODE REVIEW

File: torchtitan/trainer.py [hunk 1]
```diff
@@ -585,6 +587,10 @@ def __init__(self, config: Config):
         )
         self.fwd_bwd_fn = self._forward_backward_body
         if not config.training.disable_cuda_graphs:
+            enable_cudagraph_annotations()
```
Comment (tianyu-l) [inline torchtitan/trainer.py]: if it's always enabled, don't need this function -- just always set `with torch.cuda.graph(..., enable_annotations=True)`

File: torchtitan/distributed/cudagraph.py [hunk 2]
```diff
@@ -186,6 +191,48 @@ def cudagraph_teardown() -> None:
...
+        return
+
+    try:
+        from torch.cuda._annotate_cuda_graph_trace import annotate_trace
+    except ImportError:
+        logger.warning(
+            "torch.cuda._annotate_cuda_graph_trace not available. "
+            "Upgrade PyTorch to enable trace CUDA graph kernel annotation."
+        )
+        return
```
Comment (tianyu-l) [inline torchtitan/distributed/cudagraph.py]: don't need this code

File: torchtitan/trainer.py [hunk 3]
```diff
@@ -585,6 +587,10 @@ def __init__(self, config: Config):
...
+            config.profiler.trace_post_processor = (
```
Comment (tianyu-l) [inline torchtitan/trainer.py]: This probably should go to config, rather than being ad hoc-ly changed in trainer build. E.g. if there had been other processors, they'll be overwritten silently.

**Review by tianyu-l — CHANGES_REQUESTED** [review]

File: torchtitan/trainer.py (hunk 3 above)
Comment (SherlockNoMad) [reply torchtitan/trainer.py]: I have removed trace_post_processor as a config. it was originally added to support annotation_post_processing. since we are have enable_annotations=True, we can just always run the post_processing. 

it's a no-op when there is no annotation collected.

File: torchtitan/trainer.py (hunk 1 above)
Comment (SherlockNoMad) [reply torchtitan/trainer.py]: done

File: torchtitan/experiments/graph_trainer/tests/test_bitwise_deterministic.py [hunk 4]
```diff
@@ -400,6 +400,9 @@ def test_aot_fx_trace_vs_eager(self):
 
         self._assert_runs_match(run_eager, run_traced, "eager vs aot_fx_trace: ")
 
+    @unittest.skip(
+        "Precompiled artifacts do not provide example inputs for CUDA graph setup"
+    )
```
Comment (SherlockNoMad) [inline torchtitan/experiments/graph_trainer/tests/test_bitwise_deterministic.py:405]: will be fixed in the following PR.

File: torchtitan/tools/profiler.py [hunk 5]
```diff
@@ -310,9 +296,7 @@ def trace_handler(prof):
 
             output_file = os.path.join(curr_trace_dir, PROFILE_FILE.format(rank=rank))
             prof.export_chrome_trace(output_file)
-
-            if post_processor is not None:
-                post_processor(output_file)
+            cudagraph_annotate_trace_post_processor(output_file)
```
Comment (tianyu-l) [inline torchtitan/tools/profiler.py:299]: need to make sure this doesn't fail on non-nvidia chips

**Review by tianyu-l — APPROVED** [review]
one comment

File: torchtitan/tools/profiler.py (hunk 5 above)
Comment (SherlockNoMad) [reply torchtitan/tools/profiler.py:299]: it's a no-op when there is no annotation collected.
